Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs

```diff
@@ -81,27 +82,32 @@ public TcpSocketClient(EncryptionManager encryptionManager, bool keepAlive, bool
             }
         }
 
-        private async Task Connect(Uri uri)
+        private async Task Connect(Uri uri, TimeSpan timeOut)
         {
-            var addresses = await uri.ResolveAsync(_ipv6Enabled);
-            AggregateException innerErrors = null;
-            for (var i = 0; i < addresses.Length; i++)
+            using (CancellationTokenSource cancellationSource = new CancellationTokenSource(timeOut))
             {
-                try
-                {
-                    await _client.ConnectAsync(addresses[i], uri.Port).ConfigureAwait(false);
-                    return;
-                }
-                catch (Exception e)
+                var addresses = await uri.ResolveAsync(_ipv6Enabled);
+                AggregateException innerErrors = null;
+                for (var i = 0; i < addresses.Length; i++)
                 {
-                    var error = new IOException($"Failed to connect to server '{uri}' via IP address '{addresses[i]}': {e.Message}", e);
-                    innerErrors = innerErrors == null ? new AggregateException(error) : new AggregateException(innerErrors, error);
+                    try
+                    {
+                        cancellationSource.Token.ThrowIfCancellationRequested();
```

Comment:
I have done that change to remove Wait() call on [this line](`https://github.com/neo4j/neo4j-dotnet-driver/blob/1.5/Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs#L70) and move the timeout logic to a more manageable level - without blocking.